codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2FrameWriter.java

```diff
@@ -111,123 +112,152 @@ public ChannelFuture writeData(ChannelHandlerContext ctx, int streamId, ByteBuf
             int payloadLength = data.readableBytes() + padding + flags.getPaddingPresenceFieldLength();
             verifyPayloadLength(payloadLength);
 
-            ByteBuf out = ctx.alloc().buffer(FRAME_HEADER_LENGTH + payloadLength);
+            writeFrameHeader(ctx, payloadLength, DATA, flags, streamId, promiseAggregator.newPromise());
```

Comment:
Let me re-phrase to see if I understand what your getting at.  There is an opportunity to combine multiple buffer allocations into 1 when ever we have sequential operations that allocate a different buffer (in this case `writeFrameHeader` and `writePaddingLength`).  Is this correct?  If so I'm glad you mentioned it...I haven't yet done this for interface simplicity and separation of concerns but you are right there may be enough of a performance gain to justify investigating these cases.